uniffi_core/src/lib.rs

```diff
@@ -653,6 +653,68 @@ unsafe impl<T: Sync + Send> FfiConverter for std::sync::Arc<T> {
     }
 }
 
+/// Support for passing reference-counted trait objects via the FFI.
+///
+/// This is different than the last one, since trait objects are stored as 2 pointers (AKA a "fat"
+/// pointer): one for `T` and another for the vtable. Both `Arc<dyn T>` and `Box<dyn T>` are fat,
+/// but `Arc<Box<dyn T>>` and `Box<Arc<dyn T>>` are not.
+/// Since the foreign side is expecting a single pointer, we need one of these. We use an inner
+/// `Arc<dyn T>` so we can support that directly as a receiver.
+unsafe impl<T: Sync + Send + ?Sized> FfiConverter for Box<std::sync::Arc<T>> {
```

Comment:
This is going to be an issue for #1444, since that removes `RustType`  and always implements `FfiConverter<UT>` on the type itself.  And Rust doesn't `unsafe impl<T: Sync + Send + !Sized> FfiConverter for std::sync::Arc<T>`

One option would be to add back `RustType`  make this the one impl where `RustType != Self`.  In general, that's not going to work with proc-macros, since they can't know the right `FfiConverter` based on the type name alone.  However in this case, the proc-macros can check for the `dyn` keyword and use that in their logic.  Even then though, the user might alias the `dyn` away, for example:

```
type ButtonInterface = dyn Button;
fn get_buttons() -> Vec<Arc<ButtonInterface>> { ... }
```

Another option would be to always Box the Arc before passing it to the foreign code, but that doesn't seem great.

I think my preferred option would be to make the `FfiConverter` impl specific to the trait.  For each trait interface we see in the UDL (or export with a proc-macro eventually), we define `impl<UC> FfiConverter<UC> for dyn SpecificTrait` in the scaffolding code.  Maybe it could forward the method calls to functions in `uniffi_core` to avoid duplicating the code.